### DISCUSSION TITLE
FastAPI and middleware (python beginner)

### DISCUSSION THREAD

**Discussion by atoneveverie** [fastapi/fastapi#6693] [Questions] [answered]
### First check

* [ ] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

I am new with Python and FastAPI but I have certain experience with node js and express js.
Currently I am developing a simple REST API  which is already functional.
Now I facing a small architectural issue because initially the API wasn't minded with a user management and I have to find a way to distinguish between simple users and an administrator.

My idea now is to introduce a middleware which will check if the request comes from a normal user or an admin.
And here comes my questioning, in the world of express js, we are able to call a middleware function just before the endpoint functional code (this is a lite description of how the things are working in express). Of course we are able also to load middleware, in a more global level, by taking the application object and setting to it the necessary middleware.

The second approach of adding a middleware under express is very similar to the one described in FastAPI documentation (https://fastapi.tiangolo.com/tutorial/middleware/ & https://fastapi.tiangolo.com/advanced/middleware/). But in both cases we add this middleware in the beginning, before getting into the router.

I would like to obtain something similar to what is available in express,  call the middleware just before the endpoint. Why? For example, if an endpoint is called by a simple user but only admins are allowed, I would like to return 403 forbidden.

So while reading the documentation I saw that page: https://fastapi.tiangolo.com/advanced/custom-request-and-route/ (paragraph Custom APIRoute class in a router). It seems very close to what I am looking for but I am not quite sure if this is the solution.

Does anyone here has the experience of both technos (express and fastapi) to give her/his opinion ? 

Thanks a lot for your attention and help ! 

Best regards, 
Anton

**Comment by frankie567** [comment, marked as answer]
ExpressJS indeed relies heavily on chain of middlewares to achieve this kind of features.

However, with FastAPI, the philosophy is different. FastAPI emphasizes the use of **dependency injections** ; which are values/objects you are declaring at path, router or application level. FastAPI then is smart enough to resolve the dependency chain and give you the value you need. Read: https://fastapi.tiangolo.com/tutorial/dependencies/

For your authentication/authorization feature, you'll have a dependency that get you an authenticated user (through an API token for example). On top of that, you could wrap this in another dependency that checks if the authenticated user has a superuser flag (or whatever logic that suits you).

FastAPI has a very nice tutorial about how to setup your own authentication system: https://fastapi.tiangolo.com/tutorial/security/get-current-user Once you have that, it should be quite straightforward to add a superuser flag to the User model and create a `get_current_superuser` dependency.

**Comment by atoneveverie** [comment]
Thanks for the response @frankie567 , I believe this is exactly what I a need :)
I think, if other people are asking the same question, this might be a very good example for how the dependencies can be used: https://github.com/tiangolo/full-stack-fastapi-postgresql and more precisely https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/api .
